jbpm-case-mgmt/jbpm-case-mgmt-impl/src/main/java/org/jbpm/casemgmt/impl/CaseRuntimeDataServiceImpl.java

```diff
@@ -340,43 +340,36 @@ public CaseDefinition getCase(String deploymentId, String caseDefinitionId) {
-        
-        Collection<org.jbpm.services.api.model.NodeInstanceDesc> nodes = runtimeDataService.getNodeInstancesByCorrelationKeyNodeType(correlationKey, 
-                                                                                                Arrays.asList(ProcessInstance.STATE_ACTIVE), 
-                                                                                                Arrays.asList("MilestoneNode"), 
-                                                                                                queryContext);
-        
-        Collection<Long> completedNodes = nodes.stream().filter(n -> ((NodeInstanceDesc)n).getType() == 1).map(n -> n.getId()).collect(toList());
-        Predicate<org.jbpm.services.api.model.NodeInstanceDesc> filterNodes = null;
-        if (achievedOnly) {            
-            filterNodes = n -> ((NodeInstanceDesc)n).getType() == 1;             
-        } else {
-            filterNodes = n -> ((NodeInstanceDesc)n).getType() == 0;
-        }
-        List<String> foundMilestones = new ArrayList<>();
-        
-        List<CaseMilestoneInstance> milestones = nodes.stream()
-        .filter(filterNodes)
-        .map(n -> {
-            foundMilestones.add(n.getName());
-            return new CaseMilestoneInstanceImpl(String.valueOf(n.getId()), n.getName(), completedNodes.contains(n.getId()), n.getDataTimeStamp());        
-        })
-        .collect(toList());
-        
+
+        Collection<org.jbpm.services.api.model.NodeInstanceDesc> nodes = runtimeDataService.getNodeInstancesByCorrelationKeyNodeType(correlationKey,
+                                                                                                                                     Arrays.asList(ProcessInstance.STATE_ACTIVE, ProcessInstance.STATE_ABORTED),
+                                                                                                                                     Arrays.asList("MilestoneNode"),
+                                                                                                                                     queryContext);
+
+        // achieved milestones
+        Map<String, List<CaseMilestoneInstance>> milestonesGroup = nodes.stream()
+                                                                        .map(n -> (NodeInstanceDesc) n)
+                                                                        .filter(n -> n.getType() == 1)
+                                                                        .map(n -> new CaseMilestoneInstanceImpl(String.valueOf(n.getId()), n.getName(), true, n.getDataTimeStamp()))
+                                                                        .collect(Collectors.groupingBy(CaseMilestoneInstance::getName));
+
+        List<CaseMilestoneInstance> milestones = milestonesGroup.values().stream().map(e -> e.stream().min((o1, o2) -> o1.getAchievedAt().compareTo(o2.getAchievedAt())).get()).collect(toList());
```

Comment:
The map contains lists with one element, so ordering is applying to that list and therefore not working.
Perhaps you can order after generating the list of CaseMilestoneInstance.

`milestones.sort(Comparator.comparing(CaseMilestoneInstance::getAchievedAt));`

Reply:
Not really sure I understand your comments. If a case is cancelled and then reopen it should contain at least two elements and so on ?... do you mind to check again ?

Reply:
I meant for the failing test, which contained only one element per key, and the list obtained from the map was not ordered. Now, it's working fine, thanks!